### ISSUE TITLE
Stop silently converting map Elem from Resource to TypeString

### ISSUE THREAD

**Issue by josephholsten** [hashicorp/terraform#15327]
In (`helper/schema/schema.go: getValueType`](https://github.com/hashicorp/terraform/blob/master/helper/schema/schema.go#L1364) we see:

```go
if _, ok := schema.Elem.(*Resource); ok {
	// TODO: We don't actually support this (yet)
	// but silently pass the validation, until we decide
	// how to handle nested structures in maps
	return TypeString, nil
}
```

This has allowed us to write a provider which looks like it's actually validating its elements, but does no such thing.

```
return &schema.Resource{
	Schema: map[string]*schema.Schema{
		"create_vnic_details": {
			Type:     schema.TypeMap,
			Optional: true,
			Elem: &schema.Resource{
				Schema: map[string]*schema.Schema{
					"assign_public_ip": {
						Type:     schema.TypeBool,
						Optional: true,
					},
				},
			},
		},
	},
}
```

I do notice this was originally introduced by @radeksimko in https://github.com/hashicorp/terraform/commit/1df1c21d5b6a45f8f4cbfffa058e17165c1935b9

### Expected Behavior

`getValueType()` should have returned an error like `"create_vnic_details: unexpected map value type: schema.Resource"`

### Actual Behavior

Silently allowed using `create_vnic_details.assign_public_ip` but instead of rendering as `"true"` or `"false"`, renders as `"1"` or `"0"`

### Steps to Reproduce
I can do an actual test of this if desired. 
Create a provider with the above schema, then use to access `create_vnic_details.assign_public_ip`

**Comment by josephholsten** [comment]
oh, I'd argue the correct solution is for me to go and fix all the providers which are broken. I'm not saying we should just remove the coercion and call it a day.

**Comment by radeksimko** [comment]
I completely agree with you that we should treat such schema as invalid, in fact that's exactly what I did in https://github.com/hashicorp/terraform/pull/12638 and then I realized I broke many acceptance tests across 4 different providers - as mentioned in https://github.com/hashicorp/terraform/pull/12722

It seems that for common use cases these schemas are "kind of working" (based on passing acceptance tests) - which also scares me a bit, but hey 🤷‍♂️ 

If you are (or anyone else is) willing to modify all those schemas and tighten the validation I'm totally willing to review & merge a PR. 😉

**Comment by josephholsten** [comment]
It's comforting to know I wasn't the only one who's made this mistake!
